lib/content_based.py

```diff
@@ -67,3 +61,47 @@ def get_document_topic_distribution(self):
         :rtype: ndarray
         """
         return self.document_distribution
+
+    def get_predictions(self):
+        """
+        Get the expected ratings between users and items.
+
+        :returns: A matrix of users X documents
+        :rtype: float[][]
+        """
+        # The matrix V * VT is a (cosine) similarity matrix, where V is the row-normalized
+        # latent document matrix, this matrix is big, so we avoid having it in inline computations
+        # by changing the multiplication order
+        # predicted_rating[u,i] = sum[j]{R[u,j] Vj * Vi} / sum[j]{Vj * Vi}
+        #                       = sum[j]{R[u,j] * cos(i, j)} / sum[j]{cos(i, j)}
+        V = self.document_distribution.copy()
+        for item in range(V.shape[0]):
+            V[item] /= numpy.sqrt(V[item].dot(V[item]))
+        self.predicted_ratings = self.ratings.dot(V).dot(V.T) / V.dot(V.T.dot(numpy.ones((V.shape[0],))))
```

Comment:
having predictions as instance variable didn't solve anything, you are still recomputing for every method call. Initialize to None and only compute if None